File: src/main/java/me/melijn/jda/utils/Task.java
package me.melijn.jda.utils;

public final class Task implements Runnable {

    private final Runnable runnable;
    private final MessageHelper messageHelper;

    public Task(final MessageHelper messageHelper, final Runnable runnable) {
        this.runnable = runnable;
        this.messageHelper = messageHelper;
    }

    @Override
    public void run() {
        try {
            runnable.run();
        } catch (Exception e) {
            e.printStackTrace();
            messageHelper.printException(Thread.currentThread(), e, null, null);
        }
    }
}

File: src/main/kotlin/me/melijn/melijnbot/objects/threading/Task.kt
package me.melijn.melijnbot.objects.threading

import kotlinx.coroutines.runBlocking
import me.melijn.melijnbot.objects.utils.sendInGuild


class Task(private val func: suspend () -> Unit) : KTRunnable {

    override suspend fun run() {
        try {
            func()
        } catch (e: Throwable) {
            e.printStackTrace()
            e.sendInGuild()
        }
    }
}

class RunnableTask(private val func: suspend () -> Unit) : Runnable {

    override fun run() {
        runBlocking {
            try {
                func()
            } catch (e: Throwable) {
                e.printStackTrace()
                e.sendInGuild()
            }
        }
    }
}

class TaskInline(private inline val func: () -> Unit) : Runnable {

    override fun run() {
        try {
            func()
        } catch (e: Throwable) {
            e.printStackTrace()
            e.sendInGuild()
        }
    }
}

@FunctionalInterface
interface KTRunnable {
    suspend fun run()
}